File: luogu/UVA_11178_Morley_s_Theorem.py
from decimal import Decimal, getcontext
import math


getcontext().prec = 18
# Define EPSILON based on the current precision context
# It's set to 10^-(prec-20), ensuring the exponent is at least 1.
_epsilon_exponent = max(1, getcontext().prec - 20)
EPSILON = Decimal("1e-" + str(_epsilon_exponent))
import sys

input = sys.stdin.readline


class Vector:
    def __init__(self, x, y):
        self.x = Decimal(x)
        self.y = Decimal(y)

    def __sub__(self, other):
        if not isinstance(other, Vector):
            return NotImplemented
        return Vector(self.x - other.x, self.y - other.y)

    def __mul__(self, other):  # 标量乘法
        if isinstance(other, (int, float, Decimal)):
            return Vector(self.x * other, self.y * other)
        raise TypeError("Multiplication is only supported with a scalar.")

    def __neg__(self):
        return Vector(-self.x, -self.y)

    def cross_product(self, other):
        """计算二维叉积 (self x other)。"""
        if not isinstance(other, Vector):
            raise TypeError("Cross product is only supported between two Vectors.")
        return self.x * other.y - self.y * other.x

    def dot_product(self, other):
        """计算二维点积 (self . other)。"""
        if not isinstance(other, Vector):
            raise TypeError("Dot product is only supported between two Vectors.")
        return self.x * other.x + self.y * other.y

    def magnitude(self):
        """计算向量的模长。"""
        return (self.x**2 + self.y**2).sqrt()

    def angle(self, other):
        """计算两个向量之间的夹角（弧度）。"""
        if not isinstance(other, Vector):
            raise TypeError("Angle calculation is only supported between two Vectors.")

        mag_self = self.magnitude()
        mag_other = other.magnitude()

        if mag_self == 0 or mag_other == 0:
            raise ValueError("Cannot calculate angle with a zero vector.")

        dot_prod = self.dot_product(other)
        cos_theta = dot_prod / (mag_self * mag_other)

        cos_theta = max(Decimal("-1"), min(Decimal("1"), cos_theta))

        return Decimal(math.acos(cos_theta))

    def rotate(self, angle):
        """
        将向量绕原点旋转指定角度（弧度）。
        """
        cos_a = Decimal(math.cos(angle))
        sin_a = Decimal(math.sin(angle))
        new_x = self.x * cos_a - self.y * sin_a
        new_y = self.x * sin_a + self.y * cos_a
        return Vector(new_x, new_y)


class Point:
    def __init__(self, x, y):
        self.x = Decimal(x)
        self.y = Decimal(y)

    def __sub__(self, other):
        """两点相减得到一个向量。"""
        if not isinstance(other, Point):
            return NotImplemented
        return Vector(self.x - other.x, self.y - other.y)

    def __add__(self, other):
        """点与向量相加得到一个新点。"""
        if not isinstance(other, Vector):
            return NotImplemented
        return Point(self.x + other.x, self.y + other.y)


class Line:
    def __init__(self, p, v):
        self.p = p
        self.v = v

    def intersection(self, other):
        """
        计算两条直线的交点。
        """
        if not isinstance(other, Line):
            raise TypeError("Input other must be a Line.")

        cross_prod_v = self.v.cross_product(other.v)

        if abs(cross_prod_v) < EPSILON:
            if abs((other.p - self.p).cross_product(self.v)) < EPSILON:
                return None
            else:
                return None

        t_numerator = (other.p - self.p).cross_product(other.v)
        t = t_numerator / cross_prod_v
        intersection_point = self.p + self.v * t
        return intersection_point


for _ in range(int(input())):
    x1, y1, x2, y2, x3, y3 = map(Decimal, input().split())
    p1 = Point(x1, y1)
    p2 = Point(x2, y2)
    p3 = Point(x3, y3)

    v1 = p2 - p1
    v2 = p3 - p2
    v3 = p1 - p3

    ang1 = v1.angle(-v2)
    ang2 = v2.angle(-v3)
    ang3 = v3.angle(-v1)

    nv1_1 = v1.rotate(ang3 / 3)
    nv1_2 = v1.rotate(ang3 * 2 / 3)
    nv2_1 = v2.rotate(ang1 / 3)
    nv2_2 = v2.rotate(ang1 * 2 / 3)
    nv3_1 = v3.rotate(ang2 / 3)
    nv3_2 = v3.rotate(ang2 * 2 / 3)

    l1_1 = Line(p1, nv1_1)
    l1_2 = Line(p1, nv1_2)
    l2_1 = Line(p2, nv2_1)
    l2_2 = Line(p2, nv2_2)
    l3_1 = Line(p3, nv3_1)
    l3_2 = Line(p3, nv3_2)

    rp1 = l1_1.intersection(l2_2)
    rp2 = l2_1.intersection(l3_2)
    rp3 = l3_1.intersection(l1_2)

    print(
        *[
            i.quantize(Decimal("0.000000"))
            for i in [rp2.x, rp2.y, rp3.x, rp3.y, rp1.x, rp1.y]
        ]
    )

File: luogu/UVA_11178_Morley_s_Theorem.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <vector>
#include <algorithm>
using namespace std;

typedef long double ld;

const ld EPSILON = 1e-12;

class Vector {
public:
    ld x, y;

    Vector(ld x, ld y) : x(x), y(y) {}

    Vector operator-(const Vector& other) const {
        return Vector(x - other.x, y - other.y);
    }

    Vector operator*(ld scalar) const {
        return Vector(x * scalar, y * scalar);
    }

    Vector operator-() const {
        return Vector(-x, -y);
    }

    ld cross_product(const Vector& other) const {
        return x * other.y - y * other.x;
    }

    ld dot_product(const Vector& other) const {
        return x * other.x + y * other.y;
    }

    ld magnitude() const {
        return sqrt(x*x + y*y);
    }

    ld angle(const Vector& other) const {
        ld mag_self = magnitude();
        ld mag_other = other.magnitude();

        if (mag_self == 0 || mag_other == 0) {
            throw runtime_error("Cannot calculate angle with a zero vector.");
        }

        ld dot_prod = dot_product(other);
        ld cos_theta = dot_prod / (mag_self * mag_other);

        cos_theta = max(ld(-1.0), min(ld(1.0), cos_theta));

        return acos(cos_theta);
    }

    Vector rotate(ld angle) const {
        ld cos_a = cos(angle);
        ld sin_a = sin(angle);
        ld new_x = x * cos_a - y * sin_a;
        ld new_y = x * sin_a + y * cos_a;
        return Vector(new_x, new_y);
    }
};

class Point {
public:
    ld x, y;

    Point(ld x, ld y) : x(x), y(y) {}

    Vector operator-(const Point& other) const {
        return Vector(x - other.x, y - other.y);
    }

    Point operator+(const Vector& vec) const {
        return Point(x + vec.x, y + vec.y);
    }
};

class Line {
public:
    Point p;
    Vector v;

    Line(const Point& p, const Vector& v) : p(p), v(v) {}

    Point* intersection(const Line& other) {
        ld cross_prod_v = v.cross_product(other.v);

        if (abs(cross_prod_v) < EPSILON) {
            if (abs((other.p - p).cross_product(v)) < EPSILON) {
                return nullptr;
            } else {
                return nullptr;
            }
        }

        ld t_numerator = (other.p - p).cross_product(other.v);
        ld t = t_numerator / cross_prod_v;
        Point intersection_point = p + v * t;
        return new Point(intersection_point);
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout << fixed << setprecision(6);

    int t;
    cin >> t;

    while (t--) {
        ld x1, y1, x2, y2, x3, y3;
        cin >> x1 >> y1 >> x2 >> y2 >> x3 >> y3;

        Point p1(x1, y1);
        Point p2(x2, y2);
        Point p3(x3, y3);

        Vector v1 = p2 - p1;
        Vector v2 = p3 - p2;
        Vector v3 = p1 - p3;

        ld ang1 = v1.angle(-v2);
        ld ang2 = v2.angle(-v3);
        ld ang3 = v3.angle(-v1);

        Vector nv1_1 = v1.rotate(ang3 / 3);
        Vector nv1_2 = v1.rotate(ang3 * 2 / 3);
        Vector nv2_1 = v2.rotate(ang1 / 3);
        Vector nv2_2 = v2.rotate(ang1 * 2 / 3);
        Vector nv3_1 = v3.rotate(ang2 / 3);
        Vector nv3_2 = v3.rotate(ang2 * 2 / 3);

        Line l1_1(p1, nv1_1);
        Line l1_2(p1, nv1_2);
        Line l2_1(p2, nv2_1);
        Line l2_2(p2, nv2_2);
        Line l3_1(p3, nv3_1);
        Line l3_2(p3, nv3_2);

        Point* rp1 = l1_1.intersection(l2_2);
        Point* rp2 = l2_1.intersection(l3_2);
        Point* rp3 = l3_1.intersection(l1_2);

        cout << rp2->x << " " << rp2->y << " "
             << rp3->x << " " << rp3->y << " "
             << rp1->x << " " << rp1->y << "\n";

        delete rp1;
        delete rp2;
        delete rp3;
    }

    return 0;
}